### PR TITLE
Initial contribution of code from aspect-build/bazel_rules_template

### PR DESCRIPTION
**Pull request by alexeagle**
Upstreaming this code to the SIG, since that group is more qualified to recommend how new rulesets are formed.

### CODE REVIEW

File: mylang/defs.bzl [hunk 1]
```diff
@@ -0,0 +1 @@
+"Public API re-exports"
```
Comment (cgrindel) [inline mylang/defs.bzl:1]: Is `defs.bzl` the recommended convention? I thought it was `mylang.bzl`.

**Review by cgrindel — CHANGES_REQUESTED** [review]
Looks good. Thanks for putting it together.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+jobs:
+  test:
+    # The type of runner that the job will run on
+    runs-on: ubuntu-latest
+
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+      - uses: actions/checkout@v2
+      - uses: bazelbuild/setup-bazelisk@v1
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply .github/workflows/ci.yaml:23]: Right, you showed me that surprisingly GitHub Actions base images already have bazelisk

I'll update the README on this setup-bazelisk thing to indicate that it's optional and drop this line.
https://github.com/bazelbuild/setup-bazelisk/compare/main...alexeagle:patch-1

File: BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    languages = ["@bazel_skylib//gazelle/bzl"],
+)
+
+gazelle(
+    name = "gazelle",
+    gazelle = "gazelle_bin",
+)
+
+# This declares the release artifact users
+pkg_tar(
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply BUILD.bazel:16]: Yes, see discussion here
https://github.com/bazel-contrib/SIG-rules-authors/issues/11#issuecomment-971814025

It's both to get a stable SHA that you can predict ahead-of-time (so the WORKSPACE snippet can be updated in the release process for example) and also to ensure users don't install your tests and examples and docs

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+WORKSPACE in the integration_tests folder.
+
+To always tell Bazel to use this directory rather than some release
+artifact or a version fetched from the internet, run this from this
+directory:
+
+```sh
+OVERRIDE="--override_repository=rules_mylang=$(pwd)/rules_mylang"
+echo "build $OVERRIDE" >> ~/.bazelrc
+echo "query $OVERRIDE" >> ~/.bazelrc
```
Comment (aherrmann) [inline CONTRIBUTING.md:35]:
```suggestion
echo "build $OVERRIDE" >> ~/.bazelrc
echo "fetch $OVERRIDE" >> ~/.bazelrc
echo "query $OVERRIDE" >> ~/.bazelrc
```
Perhaps `fetch` as well? IIRC it doesn't inherit from `build`.

File: WORKSPACE [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+load(":internal_deps.bzl", "rules_mylang_internal_deps")
+
+# Fetch deps needed only locally for development
+rules_mylang_internal_deps()
+
+load("//mylang:repositories.bzl", "mylang_register_toolchains", "rules_mylang_dependencies")
+
+# Fetch our "runtime" dependencies which users need as well
+rules_mylang_dependencies()
```
Comment (aherrmann) [inline WORKSPACE:19]:
```suggestion
# Fetch dependencies which users need as well
rules_mylang_dependencies()
```
Nit-picking: I know "runtime" was quoted, but still, to me it seems clearer and less confusing to just omit it.

File: .bazelrc [hunk 6]
```diff
@@ -0,0 +1,13 @@
+# Bazel settings that apply to this repository.
+# Take care to document any settings that you expect users to apply.
+# Settings that apply only to CI are in .github/workflows/ci.bazelrc
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply .bazelrc:3]: Yes, I've used that `--config=ci` scheme in the past as well. The main reason to make another file is because in a complex CI setup, you end up calling `bazel` in lots of different places. You need to query things to discover what to build & test, someone's code requires a different configuration (test fails without `-c opt` for example).
It's easy to accidentally forget the `--config=ci` flag in one of the many invocations of bazel, or especially hard to ensure that later maintainers of the CI scripts don't omit it.

So, the pattern I usually use is to copy the ci.bazelrc file into the system or user's location at the beginning of the pipeline. This way the CI rc settings are always applied without needing to remember what flags to pass each bazel call.

To be fair, in this repo, I don't think it makes much difference which we pick. If we want to simplify things, I wouldn't even bother with a separate rc file, just use the `ci` config in the same `.bazelrc` with the rest.

File: mylang/private/toolchains_repo.bzl [hunk 7]
```diff
@@ -0,0 +1,113 @@
+"""Create a repository to hold the toolchains
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply mylang/private/toolchains_repo.bzl:1]: Yeah I agree, in using this template a lot of the work is correctly removing all the toolchain bits you don't need.

However I think if we just document what to do, it will be hard to follow how to create a toolchain, and we'll end up with a lot of "code in comments" which is harder to keep it working.
If we make a separate template, they will diverge over time and we'll double the maintenance effort.
There are fancy systems for code templating like "cookiecutter" from python-land, where you can interactively ask the user what features to enable, but I don't think the SIG has time to invest in something fancier than just GitHub templates built-in feature.

So I'm torn about what is the better way to do this.

File: BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+gazelle(
+    name = "gazelle",
+    gazelle = "gazelle_bin",
+)
+
+# This declares the release artifact users fetch.
+# You could rely on GitHub to provide an archive URL that packages the repo on the fly.
+# We create a tar for a few reasons:
+# - we can calculate the checksum ourselves to write a WORKSPACE snippet during release
+# - GitHub occasionally changes their frontend servers in a way that changes all the SHAs
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply BUILD.bazel:19]: Awesome analysis, I suppose the right thing here is to remove the pkg_tar altogether. Do we think most rulesets don't have enough tests/examples/docs that it contributes meaningfully to the size of the archive users download?

**Review by Yannic — APPROVED** [review]
Great start! I fully expect this template to evolve over time, so let's not try to be perfect right from the beginning. Feel free to address any of my nits in follow-ups.

File: .bazelrc (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexeagle) [reply .bazelrc:3]: My preference is to keep all the CI configuration adjacent, more strongly than to keep all the bazelrc's adjacent. I'm open to discussion on it of course.

File: .github/workflows/ci.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexeagle) [reply .github/workflows/ci.yaml]: Dropped it

File: .pre-commit-config.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+repos:
+  # Check formatting and lint for starlark code
+  - repo: https://github.com/keith/pre-commit-buildifier
+    rev: 4.0.1.1
+    hooks:
+      - id: buildifier
+      - id: buildifier-lint
+  # Enforce that commit messages allow for later changelog generation
+  - repo: https://github.com/commitizen-tools/commitizen
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply .pre-commit-config.yaml:17]: I'm pretty sure that commitizen should be part of the setup, so that contributors remember to indicate fix/feature/breaking change. Otherwise it's so hard to determine next release version.

It will remind you about the `--no-verify` option whenever it fails, so when I'm doing quick hacking like that I just add the option.

File: BUILD.bazel (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexeagle) [reply BUILD.bazel]: Okay got this resolved, now the release automation works without it :)
Thanks @aherrmann

File: WORKSPACE [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+mylang_register_toolchains(
+    name = "mylang1_14",
+    mylang_version = "1.14.2",
+)
+
+# For running our own unit tests
+load("@bazel_skylib//lib:unittest.bzl", "register_unittest_toolchains")
+
+register_unittest_toolchains()
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply WORKSPACE:29]: fixed

File: mylang/tests/BUILD.bazel [hunk 11]
```diff
@@ -0,0 +1,3 @@
+load(":versions_test.bzl", "versions_test_suite")
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply mylang/tests/BUILD.bazel:1]: yup, that will be a good thing to hook up when ready

File: .github/workflows/release.yml [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: bazel test //...
+        env:
+          # Bazelisk will download bazel to here
+          XDG_CACHE_HOME: ~/.cache/bazel-repo
+        run: bazel --bazelrc=.github/workflows/ci.bazelrc --bazelrc=.bazelrc test //...
+      - name: Rename release artifact with version
+        run: cp $(ls bazel-out/*/bin/*.tar.gz | tail -1) "rules_mylang-$(git describe --tags | sed 's/^v//').tar.gz"
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply .github/workflows/release.yml]: Rewrote this whole section, not needed when there's no release artifact to publish.

File: BUILD.bazel [hunk 13]
```diff
@@ -0,0 +1,46 @@
+load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
+load("@rules_pkg//:pkg.bzl", "pkg_tar")
+load("@bazel_gazelle//:def.bzl", "gazelle", "gazelle_binary")
+
+gazelle_binary(
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply BUILD.bazel:4]: The gazelle dependency isn't exposed to users. Nothing in this package is loaded by them.

(In my original formulation we wouldn't even ship this file to users, but @aherrmann convinced me that the GH archive of the repo works)

File: WORKSPACE [hunk 14]
```diff
@@ -0,0 +1,40 @@
+# Declare the local Bazel workspace.
+# This is *not* included in the published distribution.
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply WORKSPACE:2]: Bazel ignores the WORKSPACE that you include.
But, I've removed the release tarball anyhow.

File: WORKSPACE [hunk 15]
```diff
@@ -0,0 +1,40 @@
+# Declare the local Bazel workspace.
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply WORKSPACE:1]: there's a bug with WORKSPACE.bazel - I can track it down if you like. We had to give up on it for all clients so far.

File: docs/BUILD.bazel [hunk 16]
```diff
@@ -0,0 +1,13 @@
+# This load statement must be in the docs/ package rather than anything users depend on
+# so that the dependency on stardoc doesn't leak to them.
+load("@aspect_bazel_lib//lib:docs.bzl", "stardoc_with_diff_test", "update_docs")
+
+stardoc_with_diff_test(
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply docs/BUILD.bazel:5]: This rule works on windows, but I think you mean the underlying bazel-skylib diff_test being broken on windows at some recent commit?

File: mylang/private/toolchains_repo.bzl [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+and the latter will only be fetched when user actually needs to build <LANG> code.
+"
+The "complex computation" in our case is simply downloading large artifacts.
+This guidance tells us how to avoid that: we put the toolchain targets in the alias repository
+with only the toolchain attribute pointing into the platform-specific repositories.
+"""
+
+# Add more platforms as needed to mirror all the binaries
+# published by the upstream project.
+PLATFORMS = {
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply mylang/private/toolchains_repo.bzl:22]: Maybe not, but I've often needed to access it somewhere else that wants to loop over them

File: mylang/repositories.bzl [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        ],
+    )
+
+########
+# Remaining content of the file is only used to support toolchains.
+########
+_DOC = "Fetch external tools needed for mylang toolchain"
+_ATTRS = {
+    "mylang_version": attr.string(mandatory = True, values = TOOL_VERSIONS.keys()),
+    "platform": attr.string(mandatory = True, values = PLATFORMS.keys()),
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply mylang/repositories.bzl:36]: users don't typically call this rule themselves. this is inside the loop that creates each platform-specific toolchain

File: mylang/repositories.bzl [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+            "https://github.com/bazelbuild/bazel-skylib/releases/download/1.1.1/bazel-skylib-1.1.1.tar.gz",
+            "https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.1.1/bazel-skylib-1.1.1.tar.gz",
+        ],
+    )
+
+########
+# Remaining content of the file is only used to support toolchains.
+########
+_DOC = "Fetch external tools needed for mylang toolchain"
+_ATTRS = {
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply mylang/repositories.bzl:34]: I've enjoyed the order of reading the file that starts with the docstring, then the public API. it's more like reading a function in Java where you have the Javadoc, then the params, then the implementation.

But I don't care that much, can go with yours

File: mylang/private/versions.bzl [hunk 20]
```diff
@@ -0,0 +1,14 @@
+"""Mirror of release info
+
+TODO: generate this file from GitHub API"""
+
+# The integrity hashes can be computed with
+# shasum -b -a 384 [downloaded file] | awk '{ print $1 }' | xxd -r -p | base64
+TOOL_VERSIONS = {
```
[the remark this answers is not in the document]
Comment (alexeagle) [reply mylang/private/versions.bzl:7]: it's mirrored from the upstream, providing the bazel downloader with info it needs to place these in the download cache

**Comment by alexeagle** [comment]
@aherrmann @cgrindel @Yannic sorry for the delay, I finally got around to trying the pure-GH-archive approach for releases, still adds the WORKSPACE snippet to the release notes automatically just by pushing a tag.

**Comment by cgrindel** [comment]
@alexeagle Just looked at the pure-GH-archive changes. Very cool. I am going to incorporate it into my repositories, this week.

File: .bazelrc (hunk 6 above)
[the remark this answers is not in the document]
Comment (aherrmann) [reply .bazelrc:3]: > It's easy to accidentally forget the `--config=ci` flag in one of the many invocations of bazel, or especially hard to ensure that later maintainers of the CI scripts don't omit it.
> 
> So, the pattern I usually use is to copy the ci.bazelrc file into the system or user's location at the beginning of the pipeline. This way the CI rc settings are always applied without needing to remember what flags to pass each bazel call.

IIUC the same can be achieved by using `build:ci` etc. in `.bazelrc` and then writing `user.bazelrc` or `~/.bazelrc` with the following contents at the beginning of the CI pipeline.
```
build --config=ci
query --config=ci
...
```

File: mylang/repositories.bzl (hunk 18 above)
[the remark this answers is not in the document]
Comment (aherrmann) [reply mylang/repositories.bzl:36]: > Also, I think platform here is a bit of an anti-pattern. IMO, we should add+register toolchains for all platforms and let Bazel select the right toolchain.

As I understand the code that is what it does in [the loop below](https://github.com/bazel-contrib/rules-template/pull/1/files/d98e9e2a00935d75c69bf225dce68239dc7dc68e#diff-b6cbbb609fc906dae2ac0bbc9d2cecf8d2ba25dd75dfdf029a177d21486ba36eR80-R86).

Another important bit is also [documented here](https://github.com/bazel-contrib/rules-template/pull/1/files/d98e9e2a00935d75c69bf225dce68239dc7dc68e#diff-b6cbbb609fc906dae2ac0bbc9d2cecf8d2ba25dd75dfdf029a177d21486ba36eR75):
> Users can avoid this macro and do these steps themselves, if they want more control.

I.e. users should be able to provide their own toolchains with their own platform constraints.
